### PR TITLE
Add --with-twoc option

### LINKED ISSUE #15: --Werror to hfst-lexc

### PR DESCRIPTION
**Pull request by mr-martian**
This PR adds support for conditionals in monolingual templates and an option to generate a .twoc file in monolingual modules and also uses the --Werror flag when compiling .lexc files.

closes #15, #59, #63, and part of #23.

### CODE REVIEW

File: apertium-init.py [hunk 1]
```diff
@@ -87,28 +87,30 @@ def init_pair(args, email):  # type: (argparse.Namespace, str) -> Tuple[Dict[str
...
+    conditionals = []
 
     if args.analyser in ['lt', 'lttoolbox']:
         files = dict(lttoolbox_language_module_files, **any_module_files)
+        if args.with_twoc:
+            raise argparse.ArgumentError('--with-twoc can only be used in hfst modules')
     elif args.analyser == 'hfst':
         files = dict(hfst_language_module_files, **any_module_files)
+        if args.with_twoc:
+            print('adding twoc')
```
Comment (sushain97) [inline apertium-init.py]: mind removing this debug statement?

**Comment by unhammer** [comment]
aren't the files normally called `.twolc` (with an L in there)?

**Comment by mr-martian** [comment]
All the times that I've encountered it it's been .twoc with no l.

**Comment by unhammer** [comment]
OK, maybe it's an Apertium thing. The compiler is called `hfst-twolc` (the original Xerox one is `twolc` I think) and e.g. Giellatekno has `sme-phon.twolc` etc., while apertium-fin uses `.twol` just to add to the confusion :)

@ftyers care to bikeshed?

**Comment by sushain97** [comment]
@mr-martian here's the patch to fix your error handling and testing:

```
commit 0f72bea26c777e2585703ce9ae12c0fb6308caab
Author: Sushain Cherivirala <sushain@skc.name>
Date:   Tue May 7 11:06:57 2019 -0700

    Improved error handling and testing

diff --git a/apertium-init.py b/apertium-init.py
index dae93c9..359f692 100755
--- a/apertium-init.py
+++ b/apertium-init.py
@@ -65,9 +65,6 @@ def init_pair(args, email):  # type: (argparse.Namespace, str) -> Tuple[Dict[str
     else:
         conditionals = ['lttoolbox1', 'lttoolbox2']
 
-    if (args.no_prob1 and args.no_rlx1) or (args.no_prob2 and args.no_rlx2):
-        raise argparse.ArgumentError('Both --no-prob and --no-rlx can\'t be applied to the same language.')
-
     if not args.no_prob1:
         conditionals.append('prob1')
     if not args.no_prob2:
@@ -98,8 +95,6 @@ def init_lang_module(args, email):  # type: (argparse.Namespace, str) -> Tuple[D
 
     if args.analyser in ['lt', 'lttoolbox']:
         files = dict(lttoolbox_language_module_files, **any_module_files)
-        if args.with_twoc:
-            raise argparse.ArgumentError('--with-twoc can only be used in hfst modules')
     elif args.analyser == 'hfst':
         files = dict(hfst_language_module_files, **any_module_files)
         if args.with_twoc:
@@ -196,16 +191,27 @@ def main(cli_args):  # type: (List[str]) -> None
     parser.add_argument('-a1', '--analyser1', help='analyser to use for first language of pair', choices=['lt', 'lttoolbox', 'hfst'], default='lt')
     parser.add_argument('-a2', '--analyser2', help='analyser to use for second language of pair', choices=['lt', 'lttoolbox', 'hfst'], default='lt')
 
-    parser.add_argument('--no-rlx1', help='no .rlx present in first language of pair (only used for bilingual pair)', action='store_true', default=False)
-    parser.add_argument('--no-rlx2', help='no .rlx present in second language of pair (only used for bilingual pair)', action='store_true', default=False)
-    parser.add_argument('--no-prob1', help='no .prob present in first language of pair (only used for bilingual pair)', action='store_true', default=False)
-    parser.add_argument('--no-prob2', help='no .prob present in second language of pair (only used for bilingual pair)', action='store_true', default=False)
+    rlx_prob_group1 = parser.add_mutually_exclusive_group()
+    rlx_prob_group1.add_argument('--no-rlx1', help='no .rlx present in first language of pair (only used for bilingual pair)',
+                                 action='store_true', default=False)
+    rlx_prob_group1.add_argument('--no-prob1', help='no .prob present in first language of pair (only used for bilingual pair)',
+                                 action='store_true', default=False)
+
+    rlx_prob_group2 = parser.add_mutually_exclusive_group()
+    rlx_prob_group2.add_argument('--no-prob2', help='no .prob present in second language of pair (only used for bilingual pair)',
+                                 action='store_true', default=False)
+    rlx_prob_group2.add_argument('--no-rlx2', help='no .rlx present in second language of pair (only used for bilingual pair)',
+                                 action='store_true', default=False)
+
     parser.add_argument('--no-pgen1', help='no post-dix present in first language of pair (only used for bilingual pair)', action='store_true', default=False)
     parser.add_argument('--no-pgen2', help='no post-dix present in second language of pair (only used for bilingual pair)', action='store_true', default=False)
     parser.add_argument('--with-twoc', help='include .twoc file (only used for monolingual hfst modules)', action='store_true', default=False)
 
     args = parser.parse_args(cli_args)
 
+    if args.analyser in ['lt', 'lttoolbox'] and args.with_twoc:
+        parser.error('--with-twoc can only be used in hfst modules')
+
     try:
         email = subprocess.check_output(shlex.split('git config user.email')).decode('utf-8').strip()
     except subprocess.CalledProcessError as e:
diff --git a/test.py b/test.py
index d330083..c40ce00 100755
--- a/test.py
+++ b/test.py
@@ -42,14 +42,6 @@ class TestModule:
         build(self.path)
 
 
-class TestInvalidModule(unittest.TestCase):
-    name = 'eng-cat-spa'
-
-    def test_init(self):
-        with self.assertRaises(SystemExit):
-            apertium_init.main([self.name])
-
-
 class TestLttoolboxModule(TestModule, unittest.TestCase):
     @classmethod
     def setUpClass(cls):
@@ -68,6 +60,28 @@ class TestTwocModule(TestModule, unittest.TestCase):
         apertium_init.main([cls.name, '--analyser=hfst', '--with-twoc'])
 
 
+class TestInvalidTwocModule(unittest.TestCase):
+    def test_rejects(self):
+        with self.assertRaises(SystemExit):
+            apertium_init.main(['eng', '--with-twoc'])
+
+
+class TestInvalidModule(unittest.TestCase):
+    name = 'eng-cat-spa'
+
+    def test_rejects_invalid_name(self):
+        with self.assertRaises(SystemExit):
+            apertium_init.main(['eng-cat-spa'])
+
+    def test_rejects_twoc_without_hfst(self):
+        with self.assertRaises(SystemExit):
+            apertium_init.main(['eng', '--with-twoc'])
+
+    def test_no_rlx_and_no_prob(self):
+        with self.assertRaises(SystemExit):
+            apertium_init.main(['eng', '--no-rlx1', '--no-prob1'])
+
+
 class TestLttoolboxPair(TestModule, unittest.TestCase):
     name = 'eng-cat'
     path = make_path(name)
```

Waiting for @ftyers for opinions re. naming. I've been using `.twol` as @unhammer mentioned: https://github.com/apertium/apertium-stats-service/blob/2cde9217598a19d5e13ac0deb72fe872b19a7874/src/stats/mod.rs#L116-L117.

**Comment by jonorthwash** [comment]
@sushain97, `.twol` is what we use for morpho-phonology, and `.twoc` is what we use for restricting long-distance dependencies in the morphotactics.  The formalism is the same, and they both compile with `hfst-twolc`, but this is the naming convention we've been using.  I think @ftyers came up with it several years ago to deal with something minor (Kyrgyz 3rd person non-past?).  Maybe we should reconsider the naming convention?  `apertium-abc.abc-morphotactics.twol`??

**Pull request merged**